### PR TITLE
[clang] Support __typeof_unqual__ in all C modes

### LINKED ISSUE #76423: typeof_unqual outside C2X using alternate keyword
gcc has `__typeof_unqual__` as an alternate spelling for `typeof_unqual` which works in all c versions.
gcc docs reference: https://gcc.gnu.org/onlinedocs/gcc/Typeof.html

### PR DESCRIPTION
**Pull request by nathanchance**
GCC has added `__typeof_unqual__` to allow `typeof_unqual` to be used in all C modes (not just C23 and newer), similar to `__typeof__` and `typeof`.

https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=607d9d50ee44163cee621cd991600acaf78c2fee

The Linux kernel would like to start using `__typeof_unqual__` to strip type qualifiers such as address spaces from inputs to macros but cannot switch to C23 due to compiler version requirements.

Match GCC and allow `__typeof_unqual__` in all C modes.

Closes: https://github.com/llvm/llvm-project/issues/76423
Link: https://lore.kernel.org/CAFULd4YG21NdF_qNVBGDtXO6xnaYFeRPvKicB=gpgUUqYE=4jw@mail.gmail.com/

PR notes/commentary:

- I am not sure if there are any other tests I should add, I could not tell where `__typeof__` was being used normally versus being explicitly tested aside from this test.
- If there are any comments in the source code that should be updated, I am more than happy to do so, I just didn't want to be preemptive if they were not going to matter.

### CODE REVIEW

File: clang/test/Parser/c2x-typeof-ext-warns.c
Comment (nickdesaulniers) [inline clang/test/Parser/c2x-typeof-ext-warns.c:18]: consider adding a test for the non-underscore-suffixed variant as well.

File: clang/include/clang/Basic/TokenKinds.def
Comment (nickdesaulniers) [inline clang/include/clang/Basic/TokenKinds.def:689]: Why not c++?

**Review by nickdesaulniers — COMMENTED** [review]
Thanks for the patch!

File: clang/include/clang/Basic/TokenKinds.def
Comment (nathanchance) [reply clang/include/clang/Basic/TokenKinds.def:689]: No real reason other than matching GCC's behavior as documented in that commit message. Is `typeof_unqual` allowed in C++? It appears to be C only in GCC but I could not seem to find anything conclusive in Clang.

File: clang/test/Parser/c2x-typeof-ext-warns.c
Comment (nathanchance) [reply clang/test/Parser/c2x-typeof-ext-warns.c:18]: Should I add one for `__typeof` as well?

File: clang/include/clang/Basic/TokenKinds.def
Comment (AaronBallman) [reply clang/include/clang/Basic/TokenKinds.def:689]: I think it's reasonable to expose this in C++ as well since it's a conforming extension (it should Just Work™).

File: clang/test/Parser/c2x-typeof-ext-warns.c
Comment (AaronBallman) [reply clang/test/Parser/c2x-typeof-ext-warns.c:18]: No harm in adding that extra coverage as well.

**Review by AaronBallman — COMMENTED** [review]
Thank you for this; I think it's a very reasonable extension!

File: clang/include/clang/Basic/TokenKinds.def
Comment (nathanchance) [reply clang/include/clang/Basic/TokenKinds.def:689]: Done in 960cf9bd3a7252d93119f02c2600891d9d7183ad.

File: clang/test/Parser/c2x-typeof-ext-warns.c
Comment (nathanchance) [reply clang/test/Parser/c2x-typeof-ext-warns.c:18]: Done in 6186b113e9c75c808ff0a08455353d2bb8c3f50b.

**Review by AaronBallman — APPROVED** [review]
LGTM, please add a C++-specific test when landing, since this is also going into C++.

**Comment by nathanchance** [comment]
> LGTM, please add a C++-specific test when landing, since this is also going into C++.

Does 4f9e79bafdd2fb9d1414e474367241244bffe29d seem reasonable on that front? Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?

**Comment by AaronBallman** [comment]
> > LGTM, please add a C++-specific test when landing, since this is also going into C++.
> 
> Does [4f9e79b](https://github.com/llvm/llvm-project/commit/4f9e79bafdd2fb9d1414e474367241244bffe29d) seem reasonable on that front? 

Yup, thank you!

> Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?

I don't think there's a need for that (your patch doesn't change the status quo for `typeof_unqual`). That said, I will rarely complain about additional test coverage. ;-)

**Comment by nathanchance** [comment]
> > Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?
> 
> I don't think there's a need for that (your patch doesn't change the status quo for `typeof_unqual`). That said, I will rarely complain about additional test coverage. ;-)

Since it is a new test, I will add that coverage just to make it a little more obvious that is expected: c89b9694c8dc05b1416d2076a067316ed19f0049

I'll merge this once presubmit is finished and no obvious regressions are noticed, thank you both for the quick feedback!

**Review by nickdesaulniers — APPROVED** [review]

File: clang/include/clang/Basic/TokenKinds.def
Comment (frederick-vs-ja) [reply clang/include/clang/Basic/TokenKinds.def]: > No real reason other than matching GCC's behavior as documented in that commit message. Is `typeof_unqual` allowed in C++? It appears to be C only in GCC but I could not seem to find anything conclusive in Clang.

The paper for adding `typeof` and `typeof_unqual` to C++ seems rejected currently (cplusplus/papers#1630).
But I do believe `typeof_unqual` is useful in C++ as it can improve compiler throughput in cases like `std::remove_cvref_t<decltype((e))>`.

**Comment by nathanchance** [comment]
Would it make sense to apply this to `release/18.x` to get it into the hands of users quicker? It seems like it should be a safe change to apply. It sounds like GCC 14 will be the first one to support `__typeof_unqual__`, which should be released quite soon if I understand their release cycle correctly, so it makes sense that Clang would support it around the same time. No worries if it does not seem reasonable, just figured it was worth asking.

**Comment by AaronBallman** [comment]
I agree that the feature is safe, but I don't think it qualifies because it's adding a new feature rather than fixing an issue. That said, if @tru or @tstellar think it's worth adding as release managers, I won't block it (because it really should be quite safe).

File: clang/include/clang/Basic/TokenKinds.def
Comment (h-vetinari) [reply clang/include/clang/Basic/TokenKinds.def]: The C++ rejection should be questioned IMO. To my understanding, once upon a time C++ went with `decltype` to not steal `typeof` from C, waiting for WG14 to standardize it (and once that happened - admittedly decades later -, it was [explicitly](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2930.pdf) made more compatible with C++ during that process). Now that it's finally there, clang could reasonably expose this in all C/C++ modes, also without the leading underscores.

**Comment by tru** [comment]
The final decision is Toms, but I don't think it qualifies since we are so late in the current process and that 19 will start in just a few months.

**Pull request merged**
